apps/src/netsim/NetSimRouterNode.js

```diff
@@ -292,11 +358,140 @@ NetSimRouterNode.prototype.buildRow_ = function () {
+  this.simulationTime_ = clock.time;
   this.heartbeat_.tick(clock);
+  this.tickRouting_.call(this, clock);
+  if (this.dnsMode === DnsMode.AUTOMATIC) {
+    this.tickAutoDns_(clock);
+  }
+};
+
+/**
+ * @param {RunLoop.Clock} clock
+ * @private
+ */
+NetSimRouterNode.prototype.tickRouting_ = function (clock) {
+  if (this.isRouterProcessing_) {
+    return;
+  }
+
+  // Process requests that are at or past their scheduled time.
+  var dueForRouting = this.getAndUnscheduleMessagesDueForRouting(clock);
+  if (dueForRouting.length > 0) {
+    this.isRouterProcessing_ = true;
+    this.routeMessages_(dueForRouting, function () {
+      this.isRouterProcessing_ = false;
+    }.bind(this));
+  }
+};
+
+/**
+ * Destructive operation: Removes schedule entries that are at or past their
+ * scheduled time, generates the message objects for those schedule entries,
+ * and returns them.
+ * @param {RunLoop.clock} clock
+ * @returns {NetSimMessage[]}
+ */
+NetSimRouterNode.prototype.getAndUnscheduleMessagesDueForRouting = function (clock) {
+  var dueForRouting = [];
+  this.localRoutingSchedule_ = this.localRoutingSchedule_.filter(function (entry) {
+    if (clock.time >= entry.time) {
+      dueForRouting.push(new NetSimMessage(this.shard_, entry.row));
+      return false;
+    }
+    return true;
+  }.bind(this));
+  return dueForRouting;
+};
+
+/**
+ * Examine the queue for anything that should be handled by the local
+ * simulation, that hasn't been added to the schedule yet.  Schedule it after
+ * everything that appears before it in the router queue.
+ */
+NetSimRouterNode.prototype.scheduleNewPackets = function () {
+  var row, belongsToLocalSim, notScheduled;
+  for (var i = 0; i < this.routerQueueCache_.length; i++) {
+    row = this.routerQueueCache_[i];
+    belongsToLocalSim = this.localSimulationOwnsMessageRow_(row);
+    notScheduled = !this.isScheduled(row);
+
+    if (belongsToLocalSim && notScheduled) {
+      // Add up send times of all packets in queue up to this one, and including
+      // this one, to get its send time.
+      var sendTime = this.routerQueueCache_
+          .slice(0, i + 1)
+          .reduce(function (prev, cur) {
+            return prev + this.calculateProcessingDurationForMessage_(cur);
+          }.bind(this), this.simulationTime_);
+      this.scheduleRouting(row, sendTime);
+    }
+  }
+};
+
+/**
+ * @param {messageRow} row
+ * @param {number} sendTime
+ */
+NetSimRouterNode.prototype.scheduleRouting = function (row, sendTime) {
+  this.localRoutingSchedule_.push({
+    row: row,
+    time: sendTime
+  });
+};
+
+/**
+ * @param {messageRow} messageRow
+ * @returns {boolean} TRUE if the given row is represented in the local simulation
+ *          routing schedule.
+ */
+NetSimRouterNode.prototype.isScheduled = function (messageRow) {
+  return undefined !== _.find(this.localRoutingSchedule_, function (entry) {
+    return entry.row.id === messageRow.id;
+  });
```

Comment:
Could use valueOr, but that might be more confusing in this context. I feel like this might be easier to parse as _.find(...) !== undefined, but that might just be personal preference. Up to you.